File: java/src/java/com/google/mr4c/algorithm/AlgorithmData.java
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

package com.google.mr4c.algorithm;

import com.google.mr4c.dataset.Dataset;
import com.google.mr4c.keys.DataKey;
import com.google.mr4c.keys.DataKeyDimension;
import com.google.mr4c.keys.DataKeyFilter;
import com.google.mr4c.keys.Keyspace;
import com.google.mr4c.sources.DataFileSource;

import java.net.URI;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

public class AlgorithmData {

	private Properties m_config;
	private Keyspace m_keyspace;
	private Map<String,Dataset> m_inputDatasets = Collections.synchronizedMap( new HashMap<String,Dataset>() );
	private Map<String,Dataset> m_outputDatasets = Collections.synchronizedMap( new HashMap<String,Dataset>() );
	private Set<String> m_excludedInputs = Collections.synchronizedSet( new HashSet<String>() );

	public AlgorithmData(Keyspace keyspace, Properties config) {
		m_keyspace = keyspace;
		m_config = config;
	}

	public AlgorithmData() {
		this(new Keyspace(), new Properties());
	}

	public Properties getConfig() {
		return m_config;
	}

	public void setConfig(Properties config) {
		m_config = config;
	}

	public Keyspace getKeyspace() {
		return m_keyspace;
	}

	public void setKeyspace(Keyspace keyspace) {
		m_keyspace = keyspace;
	}

	public void generateKeyspaceFromInputDatasets() {
		m_keyspace.addKeys(getAllInputKeys(true));
	}

	public Set<DataKey> getAllInputKeys() {
		return getAllInputKeys(false);
	}

	private Set<DataKey> getAllInputKeys(boolean exclude) {
		Set<DataKey> keys = new HashSet<DataKey>();
		for ( String name : m_inputDatasets.keySet() ) {
			if ( exclude && m_excludedInputs.contains(name) ) {
				continue;
			}
			Dataset dataset = m_inputDatasets.get(name);
			keys.addAll(dataset.getAllFileKeys());
			keys.addAll(dataset.getAllMetadataKeys());
		}
		return keys;
	}

	public Dataset getInputDataset(String name) {
		if ( !hasInputDataset(name) ) {
			throw new IllegalArgumentException(String.format("No input dataset named [%s]", name));
		}
		return m_inputDatasets.get(name);
	}

	public boolean hasInputDataset(String name) {
		return m_inputDatasets.containsKey(name);
	}

	public Set<String> getInputDatasetNames() {
		return m_inputDatasets.keySet();
	}

	public void addInputDataset(String name, Dataset dataset) {
		if ( m_inputDatasets.containsKey(name) ) {
			throw new IllegalStateException(String.format("Already have input dataset named [%s]", name));
		}
		m_inputDatasets.put(name,dataset);
	}

	public void excludeInputDatasetFromKeyspace(String name) {
		m_excludedInputs.add(name);
	}

	public Set<String> getExcludedInputDatasets() {
		return m_excludedInputs;
	}

	public Dataset getOutputDataset(String name) {
		if ( !hasOutputDataset(name) ) {
			throw new IllegalArgumentException(String.format("No output dataset named [%s]", name));
		}
		return m_outputDatasets.get(name);
	}

	public boolean hasOutputDataset(String name) {
		return m_outputDatasets.containsKey(name);
	}

	public Set<String> getOutputDatasetNames() {
		return m_outputDatasets.keySet();
	}

	public void addOutputDataset(String name, Dataset dataset) {
		if ( m_outputDatasets.containsKey(name) ) {
			throw new IllegalStateException(String.format("Already have output dataset named [%s]", name));
		}
		m_outputDatasets.put(name,dataset);
	}

	public void releaseInputs() {
		for ( Dataset dataset : m_inputDatasets.values() ) {
			dataset.release();
		}
	}

	public void releaseOutputs() {
		for ( Dataset dataset : m_outputDatasets.values() ) {
			dataset.release();
		}
	}

	public boolean equals(Object obj) {
		if ( this==obj ) return true;
		if ( !obj.getClass().equals(this.getClass()) ) return false;
		AlgorithmData algoData = (AlgorithmData) obj;
		if ( !m_keyspace.equals(algoData.m_keyspace) ) return false;
		if ( !m_config.equals(algoData.m_config) ) return false;
		if ( !m_inputDatasets.equals(algoData.m_inputDatasets) ) return false;
		if ( !m_outputDatasets.equals(algoData.m_outputDatasets) ) return false;
		if ( !m_excludedInputs.equals(algoData.m_excludedInputs) ) return false;
		return true; 
	}

	public int hashCode() {
		// this is expensive, just doing because we overrode equals
		// this class shouldn't be used as a key to begin with
		return m_keyspace.hashCode() +
			m_config.hashCode() +
			m_inputDatasets.hashCode() +
			m_outputDatasets.hashCode();
	}

	public AlgorithmData slice(DataKeyFilter filter) {
		AlgorithmData slice = new AlgorithmData();
		for ( String excluded : m_excludedInputs ) {
			slice.excludeInputDatasetFromKeyspace(excluded);
		}
		for ( String name : getInputDatasetNames() ) {
			Dataset dataset = getInputDataset(name);
			slice.addInputDataset(name, dataset.slice(filter));
		}
		for ( String name : getOutputDatasetNames() ) {
			Dataset dataset = getOutputDataset(name);
			slice.addOutputDataset(name, dataset.slice(filter));
		}
		Properties config = slice.getConfig();
		config.putAll(getConfig());
		slice.generateKeyspaceFromInputDatasets();
		return slice;
	}

	public Set<DataKey> getDependentKeys(DataKeyDimension dim) {
		Set<DataKey> keys = new HashSet<DataKey>();
		for ( String name : getInputDatasetNames() ) {
			Dataset dataset = getInputDataset(name);
			keys.addAll(dataset.getDependentKeys(dim));
		}
		return keys;
	}

	public Collection<DataFileSource> getInputDataFileSources(boolean exclude) {
		Collection<DataFileSource> sources = new ArrayList<DataFileSource>();
		for ( String name : getInputDatasetNames() ) {
			if ( exclude && m_excludedInputs.contains(name) ) {
				continue;
			}
			Dataset dataset = getInputDataset(name);
			sources.addAll(dataset.getDataFileSources());
		}
		return sources;
	}

	public Set<DataKey> getExcludedInputDatasetKeys() {
		Set<DataKey> keys = new HashSet<DataKey>();
		for ( String name : m_excludedInputs ) {
			Dataset dataset = getInputDataset(name);
			keys.addAll(dataset.getAllFileKeys());
			keys.addAll(dataset.getAllMetadataKeys());
		}
		return keys;
	}

}

File: native/src/cpp/impl/algorithm/AlgorithmData.cpp
/**
  * Copyright 2014 Google Inc. All rights reserved.
  * 
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * 
  *     http://www.apache.org/licenses/LICENSE-2.0
  * 
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

#include <map>
#include <string>
#include <stdexcept>
#include <iostream>
#include <log4cxx/logger.h>

#include "algorithm/algorithm_api.h"
#include "dataset/dataset_api.h"
#include "keys/keys_api.h"
#include "util/util_api.h"

using log4cxx::LoggerPtr;

namespace MR4C {

class AlgorithmDataImpl {

	friend class AlgorithmData;

	private:

		LoggerPtr m_logger;
		Keyspace m_keyspace;
		AlgorithmConfig m_config;
		std::map<std::string,Dataset*> m_input;
		std::set<std::string> m_inputNames;
		std::map<std::string,Dataset*> m_output;
		std::set<std::string> m_outputNames;

		AlgorithmDataImpl() {
			init();
		}

		void init() {
			m_logger = MR4CLogging::getLogger("algorithm.AlgorithmData");
		}

		void setKeyspace(const Keyspace& keyspace) {
			m_keyspace = keyspace;
		}

		const Keyspace& getKeyspace() const {
			return m_keyspace;
		}

		void generateKeyspaceFromInputDatasets() {
			LOG4CXX_INFO(m_logger, "Generating keyspace from input datasets");
			KeyspaceBuilder builder;
			std::map<std::string,Dataset*>::iterator iter = m_input.begin();
			for ( ; iter!=m_input.end(); iter++ ) {
				Dataset* dataset = iter->second;
				builder.addKeys(dataset->getAllFileKeys());
				builder.addKeys(dataset->getAllMetadataKeys());
			}
			m_keyspace = builder.toKeyspace();
		}

		void setConfig(const AlgorithmConfig& config) {
			m_config = config;
		}
			
		const AlgorithmConfig& getConfig() const {
			return m_config;
		}
		
		void addInputDataset(const std::string& name, Dataset* dataset) {
			if ( hasInputDataset(name) ) {
				MR4C_THROW(std::invalid_argument, "Already have an input dataset named [" << name << "]");
			}
			m_input[name] = dataset;
			m_inputNames.insert(name);
			dataset->setGlobalKeyspace(m_keyspace);
			dataset->generateKeyspaces();
		}

		void addOutputDataset(const std::string& name, Dataset* dataset) {
			if ( hasOutputDataset(name) ) {
				MR4C_THROW(std::invalid_argument, "Already have an output dataset named [" << name << "]");
			}
			m_output[name] = dataset;
			m_outputNames.insert(name);
		}

		Dataset* getInputDataset(const std::string& name) const {
			if ( !hasInputDataset(name) ) {
				LOG4CXX_ERROR(m_logger, "No input dataset named [" << name <<"]");
				MR4C_THROW( std::invalid_argument, "No input dataset named [" << name <<"]");
			}
			return m_input.find(name)->second;
		}
	
		bool hasInputDataset(const std::string& name) const {
			return m_input.count(name)!=0;
		}
		
		Dataset* getOutputDataset(const std::string& name) const {
			if ( !hasOutputDataset(name) ) {
				LOG4CXX_ERROR(m_logger, "No output dataset named [" << name <<"]");
				MR4C_THROW( std::invalid_argument, "No output dataset named [" << name <<"]");
			}
			return m_output.find(name)->second;
		}

		bool hasOutputDataset(const std::string& name) const {
			return m_output.count(name)!=0;
		}
		
		std::set<std::string> getInputDatasetNames() const {
			return m_inputNames;
		}

		std::set<std::string> getOutputDatasetNames() const {
			return m_outputNames;
		}


		bool operator==(const AlgorithmDataImpl& algoData) const {
			if ( m_keyspace!=algoData.m_keyspace) return false;
			if ( m_config!=algoData.m_config) return false;
			if ( !compareMapsOfPointers(m_input, algoData.m_input) ) return false;
			if ( !compareMapsOfPointers(m_output, algoData.m_output) ) return false;
			return true;
		}

		~AlgorithmDataImpl() {
			deleteMapOfPointers(m_input);
			deleteMapOfPointers(m_output);
		} 

};

AlgorithmData::AlgorithmData() {
	m_impl = new AlgorithmDataImpl();
}

AlgorithmData::~AlgorithmData() {
	delete m_impl;
} 

void AlgorithmData::setKeyspace(const Keyspace& keyspace) {
	m_impl->setKeyspace(keyspace);
}

const Keyspace& AlgorithmData::getKeyspace() const {
	return m_impl->getKeyspace();
}

void AlgorithmData::generateKeyspaceFromInputDatasets() {
	m_impl->generateKeyspaceFromInputDatasets();
}

void AlgorithmData::setConfig(const AlgorithmConfig& config) {
	m_impl->setConfig(config);
}

const AlgorithmConfig& AlgorithmData::getConfig() const {
	return m_impl->getConfig();
}

void AlgorithmData::addInputDataset(const std::string& name, Dataset* dataset) {
	m_impl->addInputDataset(name, dataset);
}

void AlgorithmData::addOutputDataset(const std::string& name, Dataset* dataset) {
	m_impl->addOutputDataset(name, dataset);
}

Dataset* AlgorithmData::getInputDataset(const std::string& name) const {
	return m_impl->getInputDataset(name);
}

bool AlgorithmData::hasInputDataset(const std::string& name) const {
	return m_impl->hasInputDataset(name);
}

Dataset* AlgorithmData::getOutputDataset(const std::string& name) const {
	return m_impl->getOutputDataset(name);
}

bool AlgorithmData::hasOutputDataset(const std::string& name) const {
	return m_impl->hasOutputDataset(name);
}

std::set<std::string> AlgorithmData::getInputDatasetNames() const {
	return m_impl->getInputDatasetNames();
}

std::set<std::string> AlgorithmData::getOutputDatasetNames() const {
	return m_impl->getOutputDatasetNames();
}

bool AlgorithmData::operator==(const AlgorithmData& algoData) const {
	return *m_impl==*algoData.m_impl;
}

bool AlgorithmData::operator!=(const AlgorithmData& algoData) const {
	return !operator==(algoData);
}

}
